src/client/app/components/visual-group/CreateVisualGroupComponent.tsx

```diff
@@ -0,0 +1,851 @@
+                console.log('made it to else if!')
+                childGroupType = 'deepGroup';
+                childMeterType = 'deepMeter';
+            }
+        }
+
+        group.childGroups.forEach(childGroup => {
+            data.links.push({
+                'target': `group_${group.id}`,
+                'source': `group_${childGroup}`,
+                'type': 'Group-to-Group',
+                'sourceType': childGroupType
+            })
+        })
+
+        group.childMeters.forEach(meter => {
+            data.links.push({
+                'target': `group_${group.id}`,
+                'source': `meter_${meter}`,
+                'type': 'meter-to-group',
+                'sourceType': childMeterType
+            })
+        })
+    });
+
+    /* Uses Topological Sorting to sort groups into columns based on their depth*/
+    const topSortAndPlaceGroups = (groups: GroupData[]): GroupData[][] => {
+        // Build adjacency list (child -> parents relationship)
+        const adjacencyList = new Map<number, number[]>();
+        const inDegree = new Map<number, number>();
+
+        // Initialize
+        groups.forEach(group => {
+            adjacencyList.set(group.id, []);
+            inDegree.set(group.id, 0);
+        });
+
+        //For each group, add edges from child to parent
+        groups.forEach(group => {
+            group.childGroups.forEach(childGroupId => {
+                // Add edge from child to parent 
+                const childEdges = adjacencyList.get(childGroupId) || [];
+                childEdges.push(group.id);
+                adjacencyList.set(childGroupId, childEdges);
+
+                // Increment in-degree of parent
+                inDegree.set(group.id, (inDegree.get(group.id) || 0) + 1);
+            });
+        });
+
+        // Utilizing Kahn's algorithm
+        const queue: number[] = [];
+        const result: GroupData[] = [];
+
+        // Add all groups who only have meters as children( with in-degree of 0)
+        groups.forEach(group => {
+            if ((inDegree.get(group.id) || 0) === 0) {
+                queue.push(group.id);
+            }
+        });
+
+        while (queue.length > 0) {
+            const currentId = queue.shift()!;
+            const currentGroup = groups.find(g => g.id === currentId);
+            if (currentGroup) {
+                result.push(currentGroup);
+            }
+
+            // Process children of current group
+            const children = groups.filter(g => g.childGroups.includes(currentId));
+            children.forEach(child => {
+                inDegree.set(child.id, (inDegree.get(child.id) || 0) - 1);
+                if ((inDegree.get(child.id) || 0) === 0) {
+                    queue.push(child.id);
+                }
+            });
+        }
+
+        // Error check for cycles
+        if (result.length !== groups.length) {
+            throw new Error('Cycle detected in group hierarchy');
```

Comment:
What is the desired behavior from the user perspective. I'm wondering what happens with this error. Often OED uses a failure popup.